81-question-mark-chrysalis/technical_specification.md

```diff
@@ -0,0 +1,263 @@
+   * `datapack` is the EM in datapack format that conforms to the schemapack identified by `schemapack_name`
+   * `annotation` is an annotation object with information from other models held by the service
+
+   This collection is filled from two sources: incoming events from the GHGA Study Repository and outputs of workflow routes executed by the EM Transformation Service.
+
+   Note that we only store published datapacks on the collection. If there are intermediate transformed datapacks, they will only be created and kept in memory as long as they are needed for running a full transformation graph corresponding to a single original datapack.
+
+#### Transformation Configuration
+
+A transformation is configured through the models, workflows and routes collections. The current configuration of the service is read from a YAML file and stored as a global `Config` object that is later used to determine whether a configuration change has occurred.
+
+Routes define the transformation workflows to transform an input schemapack to an output schemapack. Each route specifies which workflow to apply to a given input schemapack and what output schemapack it is expected to produce.
+
+Routes also define a a graph structure where schemapacks are nodes and routes are directed edges. The source nodes of this graph are the “original schemapacks” (aka EMIMs) defined as SchemaPack. The graph must not contain any "diamonds", i.e. there must be at most one directed path between any two schemapacks in the graph. This also implies that the graph does not contain any cycles, i.e. is a directed acyclic graph (DAG). 
+
+We enforce that stronger unique-path property of the graph because "diamond" shapes indicate that there is unnecessary redundancy in the graph that should be avoided, and since it leaves no ambiguity in how datapacks are transformed.
+
+This graph structure is used to validate the transformation configuration and to determine the order in which schemapacks need to be processed during service startup and when processing incoming AnnotatedEMPacks.
+
+#### Transformation Configuration Validation
+
+Transformation configuration validation includes:
+
+1. Check the workflows and the original schemapacks referred by the routes exist in their corresponding collections.
+2. Check that the schemapacks marked as `original` are not referred as route output schemapacks.
+3. Validate the schemapacks using SchemaPack library and workflows using the metldata library.
+4. Check the graph conforms to the unique path requirement mentioned above, and thereby also that it does not contain cycles.
+5. Calculate the topological order of the nodes in the graph using an algorithm similar to Kahn's algorithm, so that the schemapacks are ordered in a way that if we process the schemapacks in that order, the previous schemapack have already been processed.
```

Comment:
```suggestion
5. Calculate the topological order of the nodes in the graph using an algorithm similar to Kahn's algorithm, so that the models are ordered in a way that if we derive the schemapacks of the models in that order, the schemapacks of the previous models have already been processed.
```